core/src/main/java/hudson/model/Job.java

```diff
@@ -679,7 +670,54 @@ public void movedTo(DirectlyModifiableTopLevelItemGroup destination, AbstractIte
         }
     }
 
-    @Override public void delete() throws IOException, InterruptedException {
+    @Override
+    public void delete() throws IOException, InterruptedException {
+        checkPermission(DELETE);
+        // if a build is in progress. Cancel it.
+        if (this instanceof Queue.Task) {
+            // clear any items in the queue so they do not get picket up
+            Queue.getInstance().cancel((Queue.Task) this);
+            // interrupt any builds in progress
+            Map<Queue.Executable, Executor> building = new LinkedHashMap<>();
+            for (Computer c : Jenkins.getInstance().getComputers()) {
+                for (Executor e : c.getOneOffExecutors()) {
+                    WorkUnit workUnit = e.getCurrentWorkUnit();
+                    if (workUnit != null && (workUnit.work == this || workUnit.work.getOwnerTask() == this)) {
+                        building.put(e.getCurrentExecutable(), e);
+                        e.interrupt();
+                    }
+                }
+                for (Executor e : c.getExecutors()) {
+                    WorkUnit workUnit = e.getCurrentWorkUnit();
+                    if (workUnit != null && (workUnit.work == this || workUnit.work.getOwnerTask() == this)) {
+                        building.put(e.getCurrentExecutable(), e);
+                        e.interrupt();
+                    }
+                }
+            }
+            if (!building.isEmpty()) {
```

Comment:
🐛 this method uses cached list of builds. If the new build gets started due to whatever reason, during the deletion cycle, these builds won't be noticed.